File: 1_PythonServer/app/schemas/project/project_database.py
from typing import Any, Literal

from pydantic import BaseModel, Field


ProjectDatabaseObjectType = Literal["table", "view", "index", "trigger"]


class ProjectDatabaseObjectResponse(BaseModel):
    name: str
    type: ProjectDatabaseObjectType
    table_name: str | None = None
    sql: str | None = None


class ProjectDatabaseOverviewResponse(BaseModel):
    project_id: str
    path: str
    file_name: str
    size_bytes: int
    tables_count: int
    views_count: int
    indexes_count: int
    triggers_count: int
    objects: list[ProjectDatabaseObjectResponse] = Field(default_factory=list)


class ProjectDatabaseCellResponse(BaseModel):
    value_type: Literal["null", "integer", "real", "text", "blob"]
    value: Any = None
    size_bytes: int | None = None


class ProjectDatabaseRowsResponse(BaseModel):
    project_id: str
    path: str
    object_name: str
    columns: list[str] = Field(default_factory=list)
    rows: list[list[ProjectDatabaseCellResponse]] = Field(default_factory=list)
    limit: int
    offset: int
    has_more: bool = False


class ProjectDatabaseColumnResponse(BaseModel):
    cid: int
    name: str
    data_type: str
    not_null: bool
    default_value: str | None = None
    primary_key: int
    hidden: int = 0


class ProjectDatabaseIndexResponse(BaseModel):
    name: str
    unique: bool
    origin: str
    partial: bool


class ProjectDatabaseForeignKeyResponse(BaseModel):
    id: int
    seq: int
    table: str
    from_column: str
    to_column: str | None = None
    on_update: str | None = None
    on_delete: str | None = None
    match: str | None = None


class ProjectDatabaseTableSchemaResponse(BaseModel):
    project_id: str
    path: str
    object_name: str
    object_type: Literal["table", "view"]
    create_sql: str | None = None
    columns: list[ProjectDatabaseColumnResponse] = Field(default_factory=list)
    indexes: list[ProjectDatabaseIndexResponse] = Field(default_factory=list)
    foreign_keys: list[ProjectDatabaseForeignKeyResponse] = Field(default_factory=list)


class ProjectDatabaseQueryRequest(BaseModel):
    path: str
    sql: str = Field(min_length=1)
    limit: int = Field(default=200, ge=1)


class ProjectDatabaseQueryResponse(BaseModel):
    project_id: str
    path: str
    sql: str
    columns: list[str] = Field(default_factory=list)
    rows: list[list[ProjectDatabaseCellResponse]] = Field(default_factory=list)
    limit: int
    truncated: bool = False

File: 2_ReactWeb/src/services/project/projectDatabase.ts
import { fetchJson } from "../http/httpClient";

export type ProjectDatabaseObjectType = "table" | "view" | "index" | "trigger";

export type ProjectDatabaseObject = {
  name: string;
  type: ProjectDatabaseObjectType;
  table_name: string | null;
  sql: string | null;
};

export type ProjectDatabaseOverview = {
  project_id: string;
  path: string;
  file_name: string;
  size_bytes: number;
  tables_count: number;
  views_count: number;
  indexes_count: number;
  triggers_count: number;
  objects: ProjectDatabaseObject[];
};

export type ProjectDatabaseCell = {
  value_type: "null" | "integer" | "real" | "text" | "blob";
  value: unknown;
  size_bytes: number | null;
};

export type ProjectDatabaseRows = {
  project_id: string;
  path: string;
  object_name: string;
  columns: string[];
  rows: ProjectDatabaseCell[][];
  limit: number;
  offset: number;
  has_more: boolean;
};

export type ProjectDatabaseColumn = {
  cid: number;
  name: string;
  data_type: string;
  not_null: boolean;
  default_value: string | null;
  primary_key: number;
  hidden: number;
};

export type ProjectDatabaseIndex = {
  name: string;
  unique: boolean;
  origin: string;
  partial: boolean;
};

export type ProjectDatabaseForeignKey = {
  id: number;
  seq: number;
  table: string;
  from_column: string;
  to_column: string | null;
  on_update: string | null;
  on_delete: string | null;
  match: string | null;
};

export type ProjectDatabaseTableSchema = {
  project_id: string;
  path: string;
  object_name: string;
  object_type: "table" | "view";
  create_sql: string | null;
  columns: ProjectDatabaseColumn[];
  indexes: ProjectDatabaseIndex[];
  foreign_keys: ProjectDatabaseForeignKey[];
};

export type ProjectDatabaseQueryResult = {
  project_id: string;
  path: string;
  sql: string;
  columns: string[];
  rows: ProjectDatabaseCell[][];
  limit: number;
  truncated: boolean;
};

export function getProjectDatabaseOverview(
  projectId: string,
  path: string,
  init?: RequestInit,
) {
  const query = new URLSearchParams({ path });
  return fetchJson<ProjectDatabaseOverview>(
    `/api/projects/${encodeURIComponent(projectId)}/databases/overview?${query.toString()}`,
    init,
  );
}

export function getProjectDatabaseTableData(
  projectId: string,
  input: {
    limit: number;
    objectName: string;
    offset: number;
    path: string;
  },
  init?: RequestInit,
) {
  const query = new URLSearchParams({
    path: input.path,
    object_name: input.objectName,
    limit: String(input.limit),
    offset: String(input.offset),
  });
  return fetchJson<ProjectDatabaseRows>(
    `/api/projects/${encodeURIComponent(projectId)}/databases/table-data?${query.toString()}`,
    init,
  );
}

export function getProjectDatabaseTableSchema(
  projectId: string,
  input: {
    objectName: string;
    path: string;
  },
  init?: RequestInit,
) {
  const query = new URLSearchParams({
    path: input.path,
    object_name: input.objectName,
  });
  return fetchJson<ProjectDatabaseTableSchema>(
    `/api/projects/${encodeURIComponent(projectId)}/databases/table-schema?${query.toString()}`,
    init,
  );
}

export function queryProjectDatabase(
  projectId: string,
  input: {
    limit: number;
    path: string;
    sql: string;
  },
  init?: RequestInit,
) {
  return fetchJson<ProjectDatabaseQueryResult>(
    `/api/projects/${encodeURIComponent(projectId)}/databases/query`,
    {
      ...init,
      method: "POST",
      body: JSON.stringify({
        path: input.path,
        sql: input.sql,
        limit: input.limit,
      }),
    },
  );
}
